pkg/providers/detect.go

```diff
@@ -0,0 +1,90 @@
+package providers
+
+import (
+	"context"
+)
+
+var _ Detector = &detector{}
+
+type detector struct {
+	providerName           string
+	fetchTokenFunc         func(ctx context.Context) (string, error)
+	fetchZoneFunc          func(ctx context.Context) (string, error)
+	fetchPublicIPv4Func    func(ctx context.Context) (string, error)
+	fetchVMEnvironmentFunc func(ctx context.Context) (string, error)
+}
+
+func New(
+	name string,
+	fetchTokenFunc func(ctx context.Context) (string, error),
+	fetchZoneFunc func(ctx context.Context) (string, error),
+	fetchPublicIPv4Func func(ctx context.Context) (string, error),
+	fetchVMEnvironmentFunc func(ctx context.Context) (string, error),
+) Detector {
+	return &detector{
+		providerName:           name,
+		fetchTokenFunc:         fetchTokenFunc,
+		fetchZoneFunc:          fetchZoneFunc,
+		fetchPublicIPv4Func:    fetchPublicIPv4Func,
+		fetchVMEnvironmentFunc: fetchVMEnvironmentFunc,
+	}
+}
+
+func (d *detector) Name() string {
+	return d.providerName
+}
+
+func (d *detector) Provider(ctx context.Context) (string, error) {
+	if d.fetchTokenFunc != nil {
+		tk, err := d.fetchTokenFunc(ctx)
+		if err != nil {
+			return "", err
+		}
+		if tk != "" {
+			return d.providerName, nil
+		}
+	}
+
+	if d.fetchZoneFunc != nil {
+		zone, err := d.fetchZoneFunc(ctx)
```

Comment:
these are provider special logic.. we should hide them.

Reply:
Yes, this is not implemented in detect package (each package implements its own). Otherwise, we need to copy and paste this detect.go to each provider, which is fine as well.